### PR TITLE
Optimize opPush and SHL

### PR DESCRIPTION
**Pull request by ignasirv**
Audit optimizations:
- https://github.com/spearbit-audits/review-polygon-zkevm-dec/issues/41
- https://github.com/spearbit-audits/review-polygon-zkevm-dec/issues/26
- Use mem align for reading from SP and storing to memory instead of SHR/SHL and MSTORE32
- https://github.com/spearbit-audits/review-polygon-zkevm-jan/issues/16
- https://github.com/spearbit-audits/review-polygon-zkevm-jan/issues/16
- https://github.com/spearbit-audits/review-polygon-zkevm-jan/pull/2#discussion_r1067319783
- Optimize opmstore and opcalldataload with mem align
+ updated counter checks

### CODE REVIEW

**Review by krlosMata — APPROVED** [review]
:smile_cat:

File: main/opcodes/calldata-returndata-code.zkasm [hunk 1]
```diff
@@ -136,47 +129,51 @@ opCALLDATACOPYloop:
...
+    $ => A          :MLOAD(SP++)
+    $ => B          :MLOAD(SP)
+    $ => A          :MEM_ALIGN_RD, MSTORE(bytesToStore)
+    E => A
+    ; Store 32 bytes from calldata to memory
+                    :CALL(offsetUtil); in: [A: offset] out: [E: offset/32, C: offset%32]
+    E               :MSTORE(calldatacopyAux1)
+    $ => A          :MLOAD(MEM:E)
+    $ => B          :MLOAD(MEM:E+1)
+    ${memAlignWR_W0(A,mem.bytesToStore,C)} => D                    ; no trust calculate W0
```
Comment (chriseth) [inline main/opcodes/calldata-returndata-code.zkasm:149]: Where are these constrained?

File: main/utils.zkasm [hunk 2]
```diff
@@ -789,39 +793,16 @@ SHLarithinit:
...
-    ; init number * 2**D = init
+    B => C                  :CALL(@exp_num + RR); out:[B: 2**RR]
+    ${A/B} => D
+    C => B
+    0 => C
     E                       :ARITH, JMP(SHLarithfinal)
 
 SHLarith0:
     0 => E
                             :JMP(SHLarithfinal)
```
Comment (chriseth) [inline main/utils.zkasm:804]: This is a "jump to next" and can be removed.

File: main/opcodes/calldata-returndata-code.zkasm (hunk 1 above)
Comment (ignasirv) [reply main/opcodes/calldata-returndata-code.zkasm:149]: At `MEM_ALIGN_WR`(line 151)

File: main/opcodes/storage-memory.zkasm [hunk 3]
```diff
@@ -46,12 +48,24 @@ opMLOAD:
...
     ; store lastMemLength for memory expansion gas cost
     32              :MSTORE(lastMemLength), CALL(saveMem); in: [lastMemOffset, lastMemLength]
+    ; store bytesToStore for memalign execution
+    B               :MSTORE(bytesToStore)
+    E => A          :CALL(offsetUtil); in: [A: offset] out: [E: offset/32, C: offset%32]
+
+    E               :MSTORE(mstoreAux)
+    $ => A          :MLOAD(MEM:E)
+    $ => B          :MLOAD(MEM:E+1)
+    ${memAlignWR_W0(A,mem.bytesToStore,C)} => D                    ; no trust calculate W0
```
[the remark this answers is not in the document]
Comment (ignasirv) [reply main/opcodes/storage-memory.zkasm:62]: At MEM_ALIGN_WR(line 64)

**Pull request merged**